ttnn/cpp/ttnn/operations/data_movement/concat/concat.hpp

```diff
@@ -0,0 +1,113 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include "tt_eager/tensor/types.hpp"
+#include "ttnn/cpp/ttnn/operations/core.hpp"
+#include "tt_eager/tt_dnn/op_library/concat/concat_op.hpp"
+
+#include <ranges>
+
+
+namespace ttnn {
+namespace operations {
+namespace data_movement {
+
+struct Concat {
+
+    // Wrapper for TTDNN
+    static inline ttnn::Tensor execute_on_worker_thread(uint8_t queue_id, const std::vector<ttnn::Tensor>& input_tensors, int dim, const std::optional<MemoryConfig>& memory_config) {
+        TT_FATAL(input_tensors.size() > 0, "ttnn.concat: expected a non-empty list of Tensors!");
+
+        const auto mem_config = memory_config.value_or(ttnn::DRAM_MEMORY_CONFIG); // should match input tensor memory config when unpopulated but causes CI errors for now
+
+        if (input_tensors.size() == 1) {
+            return ttnn::to_memory_config(input_tensors.at(0), mem_config, std::nullopt);
+        }
+
+        // TODO: Issue #8426: Add validation for ttnn.concat for sharded inputs
+        // const bool all_tensors_are_tile_layout_without_padding = std::all_of(input_tensors.begin(), input_tensors.end(),
+        // [dim](const ttnn::Tensor& input_tensor){
+        //    return input_tensor.get_layout() == ttnn::TILE_LAYOUT and not has_tile_padding(input_tensor, dim);
+        //});
+        // TT_FATAL(all_tensors_are_tile_layout_without_padding, "Not Implemented");
+
+        const ttnn::Tensor& first_tensor = input_tensors.front();
+        const int rank = first_tensor.get_shape().rank();
+
+        // Wrap dim
+        dim = dim < 0 ? rank + dim : dim;
```

Comment:
I think we have some `normalize dim` function to help with this?

Reply:
Yep, switched to that.